### PR TITLE
Add more inventory commands

### LINKED ISSUE #3195: Add cmds for stonecutter, loom...
### Feature request

Hi !

**Feature description**

Is it possible to add several commands that provide an access to the stone cutter, loom, grindstone ( and others ) interfaces such as the actual **/workbench** command.

**How the feature is useful**

This feature is just to be consistent with the fact that the **/workbench** already exists and the version 1.14 added new blocks to interact/craft with.

### PR DESCRIPTION
**Pull request by JRoy**
Closes #3195 and possibly others

### CODE REVIEW

File: Essentials/src/messages.properties [hunk 1]
```diff
@@ -21,6 +21,8 @@ antiBuildPlace=\u00a74You are not permitted to place\u00a7c {0} \u00a74here.
 antiBuildUse=\u00a74You are not permitted to use\u00a7c {0}\u00a74.
 antiochCommandDescription=A little surprise for operators.
 antiochCommandUsage=/<command> [message]
+anvilCommandDescription=Opens up an Anvil.
```
Comment (mdcfe) [inline Essentials/src/messages.properties]:
```suggestion
anvilCommandDescription=Opens up an anvil.
```

File: Essentials/src/messages.properties [hunk 2]
```diff
@@ -879,6 +891,7 @@ unlimitedItemPermission=\u00a74No permission for unlimited item \u00a7c{0}\u00a7
 unlimitedItems=\u00a76Unlimited items\:\u00a7r
 unmutedPlayer=\u00a76Player\u00a7c {0} \u00a76unmuted.
 unsafeTeleportDestination=\u00a74The teleport destination is unsafe and teleport-safety is disabled.
+unsupportedBrand=\u00a74The platform you are running this on doesn't provide the capabilities for this feature
```
Comment (mdcfe) [inline Essentials/src/messages.properties]:
```suggestion
unsupportedBrand=\u00a74The server platform you are currently running does not provide the capabilities for this feature.
```

File: providers/BaseProviders/src/net/ess3/provider/ContainerProvider.java [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+import org.bukkit.inventory.InventoryView;
+
+public interface ContainerProvider extends Provider {
+
+    InventoryView openAnvil(Player player);
+    InventoryView openCartographyTable(Player player);
+    InventoryView openGrindstone(Player player);
+    InventoryView openLoom(Player player);
+    InventoryView openSmithingTable(Player player);
+    InventoryView openStonecutter(Player player);
```
Comment (mdcfe) [inline providers/BaseProviders/src/net/ess3/provider/ContainerProvider.java:18]:
```suggestion
    InventoryView openStonecutter(Player player);

```

File: providers/PaperProvider/src/net/ess3/provider/providers/PaperContainerProvider.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+    @Override
+    public InventoryView openStonecutter(Player player) {
+        return player.openStonecutter(null, true);
+    }
+
+    @Override
+    public String getDescription() {
+        return "Paper Container Opening Provider";
+    }
```
Comment (mdcfe) [inline providers/PaperProvider/src/net/ess3/provider/providers/PaperContainerProvider.java:42]:
```suggestion
    }

```

File: Essentials/src/com/earth2me/essentials/commands/Commandstonecutter.java [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+        super("stonecutter");
+    }
+
+    @Override
+    public void run(final Server server, final User user, final String commandLabel, final String[] args) throws Exception {
+        if (ess.getContainerProvider() == null) {
+            user.sendMessage(tl("unsupportedBrand"));
+            return;
+        }
+        ess.getContainerProvider().openStonecutter(user.getBase());
```
Comment (mdcfe) [inline Essentials/src/com/earth2me/essentials/commands/Commandstonecutter.java:21]:
```suggestion

        ess.getContainerProvider().openStonecutter(user.getBase());
```

File: Essentials/src/com/earth2me/essentials/commands/Commandsmithingtable.java [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+        super("smithingtable");
+    }
+
+    @Override
+    protected void run(Server server, User user, String commandLabel, String[] args) throws Exception {
+        if (ess.getContainerProvider() == null) {
+            user.sendMessage(tl("unsupportedBrand"));
+            return;
+        }
+        ess.getContainerProvider().openSmithingTable(user.getBase());
```
Comment (mdcfe) [inline Essentials/src/com/earth2me/essentials/commands/Commandsmithingtable.java:21]:
```suggestion

        ess.getContainerProvider().openSmithingTable(user.getBase());
```

File: Essentials/src/com/earth2me/essentials/commands/Commandloom.java [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+        super("loom");
+    }
+
+    @Override
+    public void run(final Server server, final User user, final String commandLabel, final String[] args) throws Exception {
+        if (ess.getContainerProvider() == null) {
+            user.sendMessage(tl("unsupportedBrand"));
+            return;
+        }
+        ess.getContainerProvider().openLoom(user.getBase());
```
Comment (mdcfe) [inline Essentials/src/com/earth2me/essentials/commands/Commandloom.java:21]:
```suggestion

        ess.getContainerProvider().openLoom(user.getBase());
```

File: Essentials/src/com/earth2me/essentials/commands/Commandgrindstone.java [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+        super("grindstone");
+    }
+
+    @Override
+    public void run(final Server server, final User user, final String commandLabel, final String[] args) throws Exception {
+        if (ess.getContainerProvider() == null) {
+            user.sendMessage(tl("unsupportedBrand"));
+            return;
+        }
+        ess.getContainerProvider().openGrindstone(user.getBase());
```
Comment (mdcfe) [inline Essentials/src/com/earth2me/essentials/commands/Commandgrindstone.java:21]:
```suggestion

        ess.getContainerProvider().openGrindstone(user.getBase());
```

File: Essentials/src/com/earth2me/essentials/commands/Commandcartographytable.java [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+        super("cartographytable");
+    }
+
+    @Override
+    public void run(final Server server, final User user, final String commandLabel, final String[] args) throws Exception {
+        if (ess.getContainerProvider() == null) {
+            user.sendMessage(tl("unsupportedBrand"));
+            return;
+        }
+        ess.getContainerProvider().openCartographyTable(user.getBase());
```
Comment (mdcfe) [inline Essentials/src/com/earth2me/essentials/commands/Commandcartographytable.java:21]:
```suggestion

        ess.getContainerProvider().openCartographyTable(user.getBase());
```

File: Essentials/src/com/earth2me/essentials/commands/Commandanvil.java [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+        super("anvil");
+    }
+
+    @Override
+    protected void run(Server server, User user, String commandLabel, String[] args) throws Exception {
+        if (ess.getContainerProvider() == null) {
+            user.sendMessage(tl("unsupportedBrand"));
+            return;
+        }
+        ess.getContainerProvider().openAnvil(user.getBase());
```
Comment (mdcfe) [inline Essentials/src/com/earth2me/essentials/commands/Commandanvil.java:21]:
```suggestion

        ess.getContainerProvider().openAnvil(user.getBase());
```

**Review by mdcfe — CHANGES_REQUESTED** [review]

File: Essentials/src/plugin.yml [hunk 11]
```diff
@@ -69,6 +73,10 @@ commands:
     description: Set a player on fire.
     usage: /<command> <player> <seconds>
     aliases: [eburn]
+  cartographytable:
+    description: Opens up a cartography table.
+    usage: /<command>
+    aliases: [ecartographytable]
```
Comment (mdcfe) [inline Essentials/src/plugin.yml]:
```suggestion
    aliases: [ecartographytable, carttable, ecarttable]
```

File: Essentials/src/plugin.yml [hunk 12]
```diff
@@ -412,6 +428,10 @@ commands:
     description: Set the owner of a player skull
     usage: /<command> [owner]
     aliases: [eskull, playerskull, eplayerskull, head, ehead]
+  smithingtable:
+    description: Opens up a smithing table.
+    usage: /<command>
+    aliases: [esmithingtable]
```
Comment (mdcfe) [inline Essentials/src/plugin.yml]:
```suggestion
    aliases: [esmithingtable, smithtable, esmithtable]
```

**Review by mdcfe — CHANGES_REQUESTED** [review]

File: providers/BaseProviders/src/net/ess3/provider/ContainerProvider.java (hunk 3 above)
Comment (mdcfe) [inline providers/BaseProviders/src/net/ess3/provider/ContainerProvider.java]:
```suggestion
    InventoryView openAnvil(Player player);

    InventoryView openCartographyTable(Player player);

    InventoryView openGrindstone(Player player);

    InventoryView openLoom(Player player);

    InventoryView openSmithingTable(Player player);

    InventoryView openStonecutter(Player player);
```

**Review by mdcfe — CHANGES_REQUESTED** [review]

**Review by mdcfe — APPROVED** [review]

**Comment by ToonTown0909** [comment]
What are the command and permissions changes associated with this PR? This is so I may apply the updates to xeya.me.

**Comment by pop4959** [comment]
> What are the command and permissions changes associated with this PR? This is so I may apply the updates to xeya.me.

Should be the command name (as seen in `super()` in the code) prefixed with `essentials.`

So for example,
```java
    public Commandanvil() {
        super("anvil");
    }
```
should have a default permission of `essentials.anvil`. You can double check this if you want before submitting the permissions to Xeyame, but that should be correct.

**Comment by ToonTown0909** [comment]
Sounds good. I'll try my best to interpret using the information you provided and go from there, will let you know if I have any other questions. (Thank you!)

**Comment by ToonTown0909** [comment]
I do hereby confirm that the new commands and permissions in relation to this EssentialsX update have been merged into the xeya.me lists.

**Pull request merged**